js-exercises/deep-equal/deepEqual.js

```diff
@@ -0,0 +1,77 @@
+function deepEqual(firstObj, secondObj, matchDescriptors = false) {
+  if (typeof matchDescriptors !== 'boolean') {
+    throw new TypeError(`Expected third argument as a boolean, got ${typeof matchDescriptors}`);
+  }
+  if (arguments.length <= 1) {
+    throw new TypeError(`Expected at least two arguments for compare, got ${arguments.length}`);
+  }
+  return checkObjectEquality(firstObj, secondObj, matchDescriptors);
+}
+
+function hasSameType(firstObj, secondObj) {
+  return typeof firstObj === typeof secondObj;
+}
+
+function checkObjectEquality(firstObj, secondObj, matchDescriptors) {
+  if (!hasSameType(firstObj, secondObj)) return false;
+  if (typeof firstObj !== 'object') {
+    if (typeof firstObj === 'function') return firstObj.toString() === secondObj.toString();
+    return firstObj === secondObj;
+  }
+  if (hasNull(firstObj, secondObj)) return true;
+  if (!hasSameKeyLength(firstObj, secondObj)) return false;
+  let isEqual = true;
+  for (const [key, firstObjValue] of Object.entries(firstObj)) {
+    let secondObjValue = secondObj[key];
+    if (matchDescriptors) {
+      const firstObjDesc = Object.getOwnPropertyDescriptor(firstObj, key);
+      const secondObjDesc = Object.getOwnPropertyDescriptor(secondObj, key);
+      isEqual = isEqual && hasSameDescriptors(firstObjDesc, secondObjDesc);
+    }
+    isEqual = isEqual && checkObjectEquality(firstObjValue, secondObjValue, matchDescriptors);
+    if (!isEqual) break;
+  }
+  return isEqual;
+}
+
+function hasNull(...values) {
+  return values.every(el => el === null);
```

Comment:
`hasNull` signifies that out of many things, some are nulls. In this case, the function will return true when every value is `null`. A better name should be `areNulls`.

Reply:
Yes, I'll change it.